subsystem/src/blocking.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Blocking interface to subsystem.
+
+use crate::{subsystem::CallError, CallResult, Handle};
+use futures::future::BoxFuture;
+use utils::shallow_clone::ShallowClone;
+
+/// Blocking version of [subsystem::Handle].
+///
+/// This should be used sparingly as blocking calls induce non-trivial overhead. The call takes up
+/// a thread in the runtime thread pool. If there is not enough threads for all simultaneous
+/// blocking calls, a new one is spawned.
+pub struct BlockingHandle<T: ?Sized>(Handle<T>);
+
+impl<T: ?Sized> Clone for BlockingHandle<T> {
+    fn clone(&self) -> Self {
+        Self::new(self.0.clone())
+    }
+}
+
+impl<T: ?Sized> ShallowClone for BlockingHandle<T> {
+    fn shallow_clone(&self) -> Self {
+        Self::new(self.0.shallow_clone())
+    }
+}
+
+impl<T: ?Sized> BlockingHandle<T> {
+    /// A new blocking handle with a dedicated worker
+    pub fn new(handle: Handle<T>) -> Self {
+        Self(handle)
+    }
+
+    /// Get the inner handle
+    pub fn handle(&self) -> &Handle<T> {
+        &self.0
+    }
+}
+
+impl<T: 'static + Send + ?Sized> BlockingHandle<T> {
+    /// Perform given closure in a worker, passing the handle to it, get the result
+    fn with_handle<R: 'static + Send>(
+        &self,
+        func: impl 'static + Send + FnOnce(&Handle<T>) -> CallResult<R>,
+    ) -> Result<R, CallError> {
+        // Get the future associated with the function call result
+        let result = func(self.handle());
+
+        // Deal with it according to the current runtime context
+        match tokio::runtime::Handle::try_current() {
+            Ok(rt) => {
+                assert_eq!(
+                    rt.runtime_flavor(),
+                    tokio::runtime::RuntimeFlavor::MultiThread,
+                    "Only multi-threaded Tokio runtime supported by blocking subsystem handle"
+                );
+                tokio::task::block_in_place(|| result.blocking_get())
+            }
+            Err(err) => {
+                if err.is_missing_context() {
+                    result.blocking_get()
```

Comment:
Could you please explain this to me? In my understanding this error can only occur if the tokio runtime hasn't been started. Are there cases when we want to have a subsystem without starting the runtime?

Reply:
My understanding is that it also happens when called from a thread that is not currently being managed by a runtime. Which is a use case worth supporting imo. Is that incorrect?

Reply:
Yes, you are right. Perhaps it is worth to add a test for such case?..

Reply:
Added a test for this